chia/wallet/wallet_singleton_store.py

```diff
@@ -0,0 +1,281 @@
+                    bytes(record.lineage_proof),
+                    record.generation,
+                    record.timestamp,
+                ),
+            )
+
+            if record.custom_data:
+                custom_data_columns = "coin_id, field_name, value"
+                coin_id = record.coin_id()
+                for key, val in record.custom_data.items():
+                    await conn.execute(
+                        f"INSERT or REPLACE INTO singleton_data ({custom_data_columns}) VALUES(?, ?, ?)",
+                        (coin_id, key, val),
+                    )
+
+    async def get_custom_data_by_coin_id(self, coin_id: bytes) -> Optional[Dict[str, Any]]:
+        async with self.db_wrapper.reader_no_transaction() as conn:
+            rows = await conn.execute_fetchall(
+                "SELECT field_name, value FROM singleton_data WHERE coin_id = ?", (coin_id,)
+            )
+        custom_data = {}
+        for row in rows:
+            custom_data[row[0]] = row[1]
+        if custom_data:
+            return custom_data
+        else:
+            return None
+
+    async def get_records_by_singleton_id(
+        self,
+        singleton_id: bytes32,
+        min_generation: Optional[uint32] = None,
+        max_generation: Optional[uint32] = None,
+        num_results: Optional[uint32] = None,
+    ) -> List[SingletonCoinRecord]:
+        """
+        Retrieves all stored singletons for a singleton ID.
+        """
+        query_params: List[Union[bytes32, uint32]] = [singleton_id]
+        for optional_param in (min_generation, max_generation, num_results):
+            if optional_param is not None:
+                query_params.append(optional_param)
+
+        async with self.db_wrapper.reader_no_transaction() as conn:
+            rows = await conn.execute_fetchall(
+                "SELECT * from singletons WHERE singleton_id=? "
+                f"{'AND generation >=? ' if min_generation is not None else ''}"
+                f"{'AND generation <=? ' if max_generation is not None else ''}"
+                "ORDER BY generation DESC"
+                f"{' LIMIT ?' if num_results is not None else ''}",
+                tuple(query_params),
+            )
+        records = []
+        for row in rows:
+            custom_data = await self.get_custom_data_by_coin_id(row[0])
+            records.append(_row_to_singleton_record(row, custom_data))
+
+        return records
+
+    async def get_record_by_coin_id(self, coin_id: bytes32) -> Optional[SingletonCoinRecord]:
+        """
+        Check for a record with coin ID and return it if present
+        """
+
+        async with self.db_wrapper.reader_no_transaction() as conn:
+            row = await execute_fetchone(
+                conn,
+                "SELECT * FROM singletons WHERE coin_id = ?",
+                (coin_id,),
+            )
+        if row is not None:
+            custom_data = await self.get_custom_data_by_coin_id(row[0])
```

Comment:
the same comment as @trepca made applies here too, about fetching all this in a single query. Although, it probably doesn't have the same impact here as in the loop above